storage/local/storage.go

```diff
@@ -360,11 +371,28 @@ func (s *memorySeriesStorage) GetMetricForFingerprint(fp clientmodel.Fingerprint
 
 // AppendSamples implements Storage.
 func (s *memorySeriesStorage) AppendSamples(samples clientmodel.Samples) {
+	if len(samples) == 0 {
+		return
+	}
+
+	s.appendMtx.Lock()
+	defer s.appendMtx.Unlock()
+
 	for _, sample := range samples {
-		s.appendSample(sample)
+		if sample.Timestamp != s.lastTimestampAppended {
+			// Timestamp has changed. We have to wait for all
+			// appendSample to complete before proceeding.
+			s.appendWaitGroup.Wait()
+			s.lastTimestampAppended = sample.Timestamp
+		}
+
+		s.appendWaitGroup.Add(1)
+		go func(sample *clientmodel.Sample) {
```

Comment:
Ok, real-life performance wins. Just for good measure, let's A/B test the load on our heaviest Prometheus pair internally soon.